### ISSUE TITLE
False positive with CheckImplicitCasts on bitfield assignment in bugprone-misplaced-widening-cast checker

### ISSUE THREAD

**Issue by earnol** [llvm/llvm-project#197261]
The `bugprone-misplaced-widening-cast` checker with `CheckImplicitCasts=true` produces a false positive when assigning to a bitfield whose declared type is wider than the source expression type, but whose bitfield width equals the  source type width.

[Reproducer](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:17,endLineNumber:9,positionColumn:17,positionLineNumber:9,selectionStartColumn:17,selectionStartLineNumber:9,startColumn:17,startLineNumber:9),source:'typedef+unsigned+int+U32%3B%0Atypedef+unsigned+long+long+U64%3B%0A%0Atypedef+struct+%7B%0A++U64+messageSize:32%3B++//+32-bit+bitfield,+same+width+as+U32%0A%7D+Header%3B%0A%0Atypedef+struct+%7B%0A++Header+header%3B%0A%7D+Token%3B%0A%0A++void+f1(U32+size)+%7B+Token+token+%3D+%7B0%7D%3B+token.header.messageSize+%3D+size+%3C%3C+1U%3B+%7D+++//+WARNS+(false+positive)%0A++void+f2(U32+size)+%7B+Token+token+%3D+%7B0%7D%3B+token.header.messageSize+%3D+size+*+2U%3B+%7D++++//+WARNS+(false+positive)%0A++void+f3(U32+size)+%7B+Token+token+%3D+%7B0%7D%3B+token.header.messageSize+%3D+size+*+2ULL%3B+%7D++//+No+warning%0A++void+f4(U32+size)+%7B+Token+token+%3D+%7B0%7D%3B+token.header.messageSize+%3D+size+%3C%3C+1ULL%3B+%7D+//+WARNS+(false+positive)%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:54.95194802148235,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:tool,i:(args:'--checks%3D%22-*,bugprone-misplaced-widening-cast%22+--config%3D!'%7BCheckOptions:+%5B%7Bkey:+bugprone-misplaced-widening-cast.CheckImplicitCasts,+value:+true%7D%5D%7D!'',argsPanelShown:'0',compilerName:'x86-64+clang+(trunk)',editorid:1,fontScale:14,fontUsePx:'0',j:1,monacoEditorHasBeenAutoOpened:'1',monacoEditorOpen:'1',monacoStdin:'1',stdin:'',stdinPanelShown:'1',toolId:clangtidytrunk,wrap:'1'),l:'5',n:'0',o:'clang-tidy+(trunk)+x86-64+clang+(trunk)+(Editor+%231,+Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang_trunk,filters:(b:'1',binary:'1',binaryObject:'1',commentOnly:'1',debugCalls:'1',demangle:'1',directives:'1',execute:'0',intel:'0',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-Xclang+-triple+-Xclang+x86_64-unknown-linux-gnu+--analyze+-Xanalyzer+-analyzer-checker%3Dcore.UndefinedBinaryOperatorResult+-Xclang+-analyzer-config+-Xclang+crosscheck-with-z3%3Dtrue+-Xclang+-analyzer-output%3Dtext',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'x86-64+clang+3.4+(assertions)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),k:45.04805197851764,l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:45.04805197851764,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4):
```
// RUN: clang-tidy '--checks=-*,bugprone-misplaced-widening-cast' \
// RUN:     --config='{CheckOptions: [{key: bugprone-misplaced-widening-cast.CheckImplicitCasts, value: true}]}' \
// RUN:     bif-field-test.c -- --target=x86_64-unknown-linux-gnu

typedef unsigned int U32;
typedef unsigned long long U64;

typedef struct {
  U64 messageSize:32;  // 32-bit bitfield, same width as U32
} Header;

typedef struct {
  Header header;
} Token;

  void f1(U32 size) { Token token = {0}; token.header.messageSize = size << 1U; }   // WARNS (false positive)
  void f2(U32 size) { Token token = {0}; token.header.messageSize = size * 2U; }    // WARNS (false positive)
  void f3(U32 size) { Token token = {0}; token.header.messageSize = size * 2ULL; }  // No warning
  void f4(U32 size) { Token token = {0}; token.header.messageSize = size << 1ULL; } // WARNS (false positive)
```
No warnings expected. The bitfield messageSize is 32 bits wide (:32), same as U32 (unsigned int). No widening occurs on assignment.

**Actual result**:
bit-field-test.c:16:49: warning: either cast from 'U32' (aka 'unsigned int') to 'U64' (aka 'unsigned long long') is ineffective, or there is loss of precision before the conversion [bugprone-misplaced-widening-cast]
